tensorflow/python/ops/custom_gradient.py

```diff
@@ -165,6 +165,73 @@ def second_order_and_transpose(ddy):
   Additional arguments to the inner `@tf.custom_gradient`-decorated function
   control the expected return values of the innermost function.
 
+  The examples above illustrate how to specify custom gradients for functions
+  which do not read from variables. The following example uses variables, which
+  require special handling because they are effectively inputs of the forward
+  function.
+
+  ```python
+  >>> weights = tf.Variable(tf.ones([2]))  # Trainable variable weights
+  >>> @tf.custom_gradient
+  ... def linear_poly(x):
+  ...   # Creating polynomial
+  ...   poly = weights[1] * x + weights[0]
+  ...
+  ...   def grad_fn(dpoly, variables=None):
+  ...     # dy/dx = weights[1] and we need to left multiply dpoly
+  ...     grad_xs = dpoly * weights[1]  # Scalar gradient
+  ...
+  ...     input_shape = len(x.shape)  # x contains how many dimensions
+  ...     grad_vars = []  # To store gradients of passed variables
+  ...     if variables is not None:
+  ...       for v in variables:
+  ...         if v is weights:
+  ...           # Manually computing dy/dweights
+  ...           dy_dw = tf.stack([x ** 1, x ** 0])
+  ...           if input_shape > 0:
+  ...             # Note here that reduce operation is performed
+  ...             # in axis so that it computes gradient of specific
+  ...             # example in batch and it doesn't reduce the entire batch
+  ...             # that is why we are only reducing on axis 1.
+  ...             grad_vars.append(
+  ...                 tf.reduce_sum(dy_dw, axis=1)
+  ...             )
+  ...           else:
+  ...             # No need to reduce if x is scalar
+  ...             grad_vars.append(dy_dw)
+  ...         else:  # Create elif if you have more variables
+  ...           grad_vars.append(None)
+  ...     return grad_xs, grad_vars
+  ...   return poly, grad_fn
+  >>> x = tf.constant([1., 2., 3.])
+  >>> with tf.GradientTape(persistent=True) as tape:
+  ...   tape.watch(x)
+  ...   poly = linear_poly(x)
+  >>> poly # poly = x + 1
+  <tf.Tensor: shape=(3,),
+    dtype=float32,
+    numpy=array([2., 3., 4.], dtype=float32)>
+  >>> tape.gradient(poly, x)  # conventional scalar gradient dy/dx
+  <tf.Tensor: shape=(3,),
+    dtype=float32,
+    numpy=array([1., 1., 1.], dtype=float32)>
+  >>> tape.gradient(poly, weights)
+  <tf.Tensor: shape=(2,),
+    dtype=float32,
+    numpy=array([6., 3.], dtype=float32)>
+  ```
+
+  Above example illustrates two things:
+  - Usage of trainable variable `weights`.
+  - Computing its gradient.
```

Comment:
I would just take this summary out, I don't think it's adding much.

Reply:
done